mythril/laser/ethereum/state.py

```diff
@@ -1,16 +1,82 @@
-from z3 import BitVec, BitVecVal, Solver, ExprRef, sat
+from z3 import BitVec, BitVecVal, BitVecRef, BitVecNumRef, BitVecSort, Solver, ExprRef, Concat, sat, simplify, Array, ForAll, Solver, UGT, Implies
+from z3.z3types import Z3Exception
 from mythril.disassembler.disassembly import Disassembly
 from copy import copy, deepcopy
 from enum import Enum
 from random import randint
+from mythril.laser.ethereum.util import get_concrete_int
 
 from mythril.laser.ethereum.evm_exceptions import StackOverflowException, StackUnderflowException
 
-
 class CalldataType(Enum):
     CONCRETE = 1
     SYMBOLIC = 2
 
+class Calldata:
+
+    def __init__(self, tx_id, starting_calldata: bytes=None):
+        self.tx_id = tx_id
+        if starting_calldata:
+            self._calldata = []
+            self.calldatasize = BitVecVal(len(starting_calldata), 256)
+            self.concrete = True
+        else:
+            self._calldata = Array('{}_calldata'.format(self.tx_id), BitVecSort(256), BitVecSort(8))
+            self.calldatasize = BitVec('{}_calldatasize'.format(self.tx_id), 256)
+            self.concrete = False
+
+        self.starting_calldata = starting_calldata if starting_calldata else []
+
+    def set_global_state(self, state):
```

Comment:
What you want to do is add constraints to the global_state right?
However, by building it like this the calldata becomes aware/dependent on GlobalState. 
Can you maybe build it where Calldata provides a function:
```python3
@property
def constraints(self):
    constraints = []
    ...
    if ...:
        x = BitVec('x', 256)
        constraints.append(ForAll(x, Implies(self[x] != 0, UGT(self.calldatasize, x))))
    return constraints
```

And the caller (who already is dependent on GlobalState) does:
globalstate.mstate.constraints.extend(calldata.constraints)

That way the calldata code is simpler and not dependent anymore

Reply:
Ah, that's what I was looking for when coding this. No clue how I missed this much better option.